aten/src/ATen/native/UnaryOps.cpp

```diff
@@ -53,17 +53,29 @@ static inline Tensor unary_op_impl(const Tensor& self, OutImpl& out_impl) {
   return out_impl(result, self);
 }
 
+// This helper function ensures the output type of the unary op is a floating point value.
+// Note that this function is only appropriate for functions like sin() which (mathematically speaking)
+// produces non-integral values. Implementers should continue to use the other unary functions, like unary_op_impl_,
+// since they explicitly specify the output type.
+template <typename OutImpl>
+static inline Tensor unary_float_op_impl(const Tensor& self, OutImpl& out_impl) {
+  Tensor result = isIntegralType(self.scalar_type(), /*include_bool=*/ true) ? \
+                  at::empty({0}, self.options().dtype(typeMetaToScalarType(c10::get_default_dtype()))) : \
+                  at::empty({0}, self.options());
+  return out_impl(result, self.to(result.scalar_type()));
+}
+
 template <typename OutImpl>
 static inline Tensor& unary_op_impl_(Tensor& self, OutImpl& out_impl) {
   return out_impl(self, self);
 }
 
 Tensor& acos_out(Tensor& result, const Tensor& self) { return unary_op_impl_out(result, self, acos_stub); }
-Tensor acos(const Tensor& self) { return unary_op_impl(self, at::acos_out); }
+Tensor acos(const Tensor& self) { return unary_float_op_impl(self, at::acos_out); }
```

Comment:
we should be able to handle the `_out` variants too.

Reply:
We can also split out `unary_op_impl_out` with a separate float-specific case.  This may make sense because that's the point where the `TensorIterator::unary_op` is constructed, so that's the point where we need to opt into or not opt into the PROMOTE behavior.

For what it's worth, I think `out=...` API calls are more closely related to the proposed `dtype=...` API calls than to the implicit promotion.  For the `out=...` and `dtype=...` variants, the user has requested that the result be a specific type.  Also, the `out=...`ness of an op variant is orthogonal to whether or not that op falls onto the implicit promote list.

Maybe it makes sense to have a single `unary_op_impl_out` with an optional `allow_promote=False` argument to serve all cases.  It isn't necessary for the present PR though. Do you think that should be part of this PR or a followup PR which adds explicit dtype flags?

Reply:
I think it makes sense to include in this PR, it's more closely related to in-place, and I don't think actually affects the mixed-precision behavior you'll want to control with `dtype=`. 

The reason this seems so relevant to this PR is in `unary_float_op_impl` the `self.to(result.scalar_type())` will be incompatible with the out variants. e.g. if a user can:

```
a=torch.tensor(1, dtype=torch.half, device='cuda')
torch.log(a, out=a)
b=torch.tensor(1, dtype=torch.int)
torch.log(b)
```
Then they should also be able to:
```
torch.log(b, out=a)
```
If we disallow the last case I think we're exposing an inconsistent API. The last case (inside TensorIterator) will validate that the result of torch.log(int_tensor) can be cast to half (by checking torch.can_cast). By casting `self` to `result.scalar_type()` explicitly here we're bypassing the code that's supposed to handle this.

Reply:
note that once implemented, an op like `torch.log(int_tensor, out=half_tensor)` will just be equivalent to (inside TensorIterator):
```
result=torch.log(int_tensor)
half_tensor.copy_(result)
```
Modifying the behavior to actually perform the operation at another precision (whether specified through `dtype=` or not) would still be a subsequent change.

Reply:
After discussion with @mruberry we're going to try option 1 first (allowing the `out=...` calls to opt-into PROMOTE).  @mruberry is that right or am i totally misinterpreting...

Reply:
Update: So we went ahead with the Option 1, after discussion with @nairbv, @mcarilli and @mruberry.

We allow `PROMOTE` for `out=...` calls as well as for unary floating ufuncs to avoid `self.to()` calls. For other cases, `result.scalar_type() != self.scalar_type()` condition in `unary_op_impl_out` definition will always return false, hence not allowing the iterator to opt-in to `PROMOTE` strategy.